compiler/x/codegen/OMRTreeEvaluator.cpp

```diff
@@ -3216,6 +3215,252 @@ TR::Register *OMR::X86::TreeEvaluator::arraycopyEvaluator(TR::Node *node, TR::Co
+   //   LEA  ESI,[ESI+8*EAX]
+   //   LOOP loopLabel
+   // endLable:
+   //   # LOOP END
+   generateRegRegInstruction(SUBRegReg(), node, dstReg, srcReg, cg);
+   if (node->isForwardArrayCopy())
+      {
+      generateRegImmInstruction(MOVRegImm4(), node, scrach, 1, cg);
+      }
+   else // decide direction during runtime
+      {
+      generateRegRegInstruction(CMPRegReg(), node, dstReg, sizeReg, cg);
+      generateRegMemInstruction(LEARegMem(), node, scrach, generateX86MemoryReference(srcReg, sizeReg, 0, -8, cg), cg);
+      generateRegRegInstruction(CMOVBRegReg(), node, srcReg, scrach, cg);
+      generateRegInstruction(SETAE1Reg, node, scrach, cg);
+      generateRegRegInstruction(MOVZXReg4Reg1, node, scrach, scrach, cg);
+      generateRegMemInstruction(LEARegMem(), node, scrach, generateX86MemoryReference(NULL, scrach, 1, -1, cg), cg);
+      }
+
+   generateRegImmInstruction(SHRRegImm1(), node, sizeReg, 3, cg);
+   generateLabelInstruction(JRCXZ1, node, endLabel, cg);
+   generateLabelInstruction(LABEL, node, loopLabel, cg);
+   generateRegMemInstruction(MOVQRegMem, node, XMM, generateX86MemoryReference(srcReg, 0, cg), cg);
+   generateMemRegInstruction(MOVQMemReg, node, generateX86MemoryReference(dstReg, srcReg, 0, cg), XMM, cg);
+   generateRegMemInstruction(LEARegMem(), node, srcReg, generateX86MemoryReference(srcReg, scrach, 3, cg), cg);
+   generateLabelInstruction(LOOP1, node, loopLabel, cg);
+
+   generateLabelInstruction(LABEL, node, endLabel, dependencies, cg);
+   cg->stopUsingRegister(XMM);
+   cg->stopUsingRegister(scrach);
+   }
+
+static void arrayCopyDefault(TR::Node* node, uint8_t elementSize, TR::Register* dstReg, TR::Register* srcReg, TR::Register* sizeReg, TR::CodeGenerator* cg)
+   {
+   TR::RegisterDependencyConditions* dependencies = generateRegisterDependencyConditions((uint8_t)3, (uint8_t)3, cg);
+   dependencies->addPreCondition(srcReg, TR::RealRegister::esi, cg);
+   dependencies->addPreCondition(dstReg, TR::RealRegister::edi, cg);
+   dependencies->addPreCondition(sizeReg, TR::RealRegister::ecx, cg);
+   dependencies->addPostCondition(srcReg, TR::RealRegister::esi, cg);
+   dependencies->addPostCondition(dstReg, TR::RealRegister::edi, cg);
+   dependencies->addPostCondition(sizeReg, TR::RealRegister::ecx, cg);
+
+   TR_X86OpCodes REPMOVS;
+   switch (elementSize)
+      {
+      case 8:
+         REPMOVS = REPMOVSQ;
+         break;
+      case 4:
+         REPMOVS = REPMOVSD;
+         break;
+      case 2:
+         REPMOVS = REPMOVSW;
+         break;
+      default:
+         REPMOVS = REPMOVSB;
+         break;
+      }
+   if (node->isForwardArrayCopy())
+      {
+      generateRepMovsInstruction(REPMOVS, node, sizeReg, dependencies, cg);
+      }
+   else // decide direction during runtime
+      {
+      TR::LabelSymbol* MainBegLabel = generateLabelSymbol(cg);
+      TR::LabelSymbol* MainEndLabel = generateLabelSymbol(cg);
+      TR::LabelSymbol* BackwardLabel = generateLabelSymbol(cg);
```

Comment:
Please begin these autos with lowercase for consistency with other vars in the code generator.